constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/LinearProfile.java

```diff
@@ -1,203 +1,134 @@
-        changePoints = changePoints.set(prev.changePoints());
-      }
-      while (iter.hasNext()) {
-        final var curr = iter.next();
-        if(!profileOutsideBounds(curr, bounds)) {
-          changePoints = changePoints.set(curr.changePoints());
-        }
-
-        if (Interval.meets(prev.interval, curr.interval)) {
-          if (prev.finalValue() != curr.initialValue && !profileOutsideBounds(prev, bounds))
-            changePoints = changePoints.set(Interval.at(prev.interval.end), true);
-        } else {
-          throw new Error("Unexpected gap in profile pieces not allowed");
-        }
-
-        prev = curr;
-      }
-
-      return changePoints.select(bounds);
-    }
-
-  /**
-   * Process each pair of intersecting profile pieces to build a list of results
-   *
-   * ASSUMPTION: Both sets of profile pieces are ordered in increasing order
-   * @param right LinearProfile to intersect with
-   * @param processor BiFunction taking two profile pieces and a desired result based on their intersection
-   * @return Set of all windows within bounds for which condition is satisfied between this and another profile
-   */
-  private static <T> List<T> processIntersections(final LinearProfile left, final LinearProfile right, final BiFunction<LinearProfilePiece, LinearProfilePiece, T> processor, Interval bounds) {
-    if (left.profilePieces.isEmpty() || right.profilePieces.isEmpty()) return new ArrayList<>();
-
-    // Setup to step through profiles simultaneously,
-    // finding all windows satisfying condition within the supplied bounds
-    final var processedIntersections = new ArrayList<T>();
-    final var rightIter = right.profilePieces.iterator();
-
-    // Loop through the left pieces, stepping up the right pieces
-    // appropriately each time to get intersections
-    if (!rightIter.hasNext()) return processedIntersections;
-    var rightPiece = rightIter.next();
-    for (final var leftPiece : left.profilePieces) {
-      //if left piece ends before bounds start, skip left piece
-      if (profileOutsideBounds(leftPiece, bounds)) continue;
-      // If left piece ends before right piece, skip left piece
-      if (Interval.compareEndToStart(leftPiece.interval, rightPiece.interval) < 0) continue;
-
-      // Step through right pieces ending before left piece starts
-      // If no right pieces ending after left piece starts exist, end the loop
-      while (Interval.compareEndToStart(rightPiece.interval, leftPiece.interval) < 0) {
-        if (rightIter.hasNext()) {
-          rightPiece = rightIter.next();
+    public Windows changePoints() {
+      final var result = IntervalMap.<Boolean>builder().set(this.profilePieces.map(LinearEquation::changing));
+      for (int i = 0; i < this.profilePieces.size(); i++) {
+        final var segment = this.profilePieces.get(i);
+        final var startTime = segment.interval().start;
+        if (i == 0) {
+          if (!segment.interval().contains(Duration.MIN_VALUE)) {
```

Comment:
Could you help me understand why we need to check for `Duration.MIN_VALUE` here? This may underlie my lack of understanding for what exactly `changePoints` means for a linear profile in the presence of gaps.

Reply:
I guess we don't have a consensus about what `Duration.MIN_VALUE` and `Duration.MAX_VALUE` mean conceptually. I see two options

1. The extrema refer to specific times, so no segments can extend beyond MIN and MAX.
2. Any segment that touches MIN and MAX is probably _intended_ to be unbounded, because realistically no bounded segment is going to extend that far. We can't actually represent unbounded yet (that's a future improvement I want to do)

I prefer the second interpretation, for that reason and because `shiftBy` under interpretation 1 would lead to unwanted gaps around the extremes.